### PR TITLE
Add's Feature Request #147, Remove Song From Playlist

### PR DESCRIPTION
**Pull request by MPierre9**
This adds the feature of removing a song from a playlist requested in issue #147.  

The code allows you to remove any song from the playlist until you reach the last index. Feedback is welcome!

aplayer-removesong

### CODE REVIEW

File: src/APlayer.js [hunk 1]
```diff
@@ -850,7 +850,59 @@ class APlayer {
             }
         }
     }
-
+    /**
+     * Remove song from playlist 
+     */
+    removeSong(indexOfSong) {
+        
+                indexOfSong = indexOfSong - 1;
```
Comment (DIYgod) [inline src/APlayer.js]: `index` should start at 0

Comment (DIYgod) [inline src/APlayer.js]: `index` should start at 0

File: src/APlayer.js [hunk 2]
```diff
@@ -850,7 +850,59 @@ class APlayer {
...
+        
+                if (this.option.music[indexOfSong] != null) { // Check if song exists 
+                    console.info("Removing song(" + indexOfSong + ")");
```
Comment (DIYgod) [inline src/APlayer.js]: useless `console.info`

File: src/APlayer.js [hunk 3]
```diff
@@ -850,7 +850,59 @@ class APlayer {
...
+                        if (oList.childNodes[i].tagName == 'LI') {
+                            liList.push(i); //Adds the LI tag indexes to array
+                        }
+                    }
+        
+                    if (this.option.music[indexOfSong + 1] != null || this.option.music[indexOfSong - 1] != null) {
+                        if (this.option.music[indexOfSong + 1] != null) { // Play next song if it exists
+                            this.pause();
+                            this.setMusic(indexOfSong + 1);
+                            this.play();
```
Comment (DIYgod) [inline src/APlayer.js]: Should keep paused if player is paused before remove song.

File: src/APlayer.js [hunk 4]
```diff
@@ -850,7 +850,59 @@ class APlayer {
...
+                    for (let i = 0; i < oList.childNodes.length; i++) {
+                        if (oList.childNodes[i].tagName == 'LI') {
+                            liList.push(i); //Adds the LI tag indexes to array
+                        }
+                    }
+        
+                    if (this.option.music[indexOfSong + 1] != null || this.option.music[indexOfSong - 1] != null) {
+                        if (this.option.music[indexOfSong + 1] != null) { // Play next song if it exists
+                            this.pause();
+                            this.setMusic(indexOfSong + 1);
```
Comment (DIYgod) [inline src/APlayer.js]: There is no need to reset if current song is not removed.

File: src/APlayer.min.js.map [hunk 5]
```diff
@@ -0,0 +1 @@
..., percentage, 'width');\n                });\n\n                // audio download error: an error occurs\n                this.audio.addEventListener('error', function () {\n                    _this2.element.getElementsByClassName('aplayer-author')[0].innerHTML = \" - Error happens \\u2565\\uFE4F\\u2565\";\n                    _this2.trigger('pause');\n                });\n\n                // audio can play: enough data is available that the media can be played\n                this.audio.addEventListener('canplay', function () {\n                    _this2.trigger('canplay');\n                });\n\n                // multiple music play\n                this.ended = false;\n                this.audio.addEventListener('ended', function () {\n                    if (_this2.multiple) {\n                        if (_this2.audio.currentTime !== 0) {\n                            if (_this2.mode === 'random') {\n                                _this2.setMusic(_this2.nextRandomNum());\n                            } else if (_this2.mode === 'single') {\n                                _this2.setMusic(_this2.playIndex);\n                            } else if (_this2.mode === 'order') {\n
```
Comment (DIYgod) [inline src/APlayer.min.js.map]: redundant file

**Comment by DIYgod** [comment]
Thanks for your request, but we need some small improvement before merged.

**Comment by MPierre9** [comment]
@DIYgod Thank you for reviewing my code :) I've made some changes let me know what you think.

**Comment by MPierre9** [comment]
Are the changes good for merge? @DIYgod

**Comment by DIYgod** [comment]
1. set to last song then pause it and then remove it will cause some errors.

1. remove a paused song will cause this player start play.

There are two more bugs I found for now.

**Comment by DIYgod** [comment]
`remove a paused song will cause this player start play` maybe caused by `setMusic `, not your fault.

**Comment by MPierre9** [comment]
Thanks @DIYgod for reviewing my code. Not sure how I missed those bugs however I revamped the code to address the bugs you outlined as well as added some functionality so APlayer will collapse from `'aplayer-withlist'` to normal when there is only one song remaining on the list. Please let me know if you find any additional bugs or features that are missing.

## Result

aplayerremovesongv2

**Pull request closed without merging**